chiselFrontend/src/main/scala/chisel3/core/MultiClock.scala

```diff
@@ -0,0 +1,25 @@
+// See LICENSE for license details.
+
+package chisel3.core
+
+import scala.language.experimental.macros
+
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl._
+import chisel3.internal.sourceinfo.{SourceInfo}
+
+private[chisel3] final case class ClockAndReset(clock: Clock, reset: Bool)
+
+object withClockAndReset {  // scalastyle:ignore object.name
+  def apply[T](clock: Clock, reset: Bool)(block: => T)(implicit sourceInfo: SourceInfo): T = {
+    // Save parentScope
+    val parentScope = Builder.currentClockAndReset
+    Builder.currentClockAndReset = Some(ClockAndReset(clock, reset))
+    val res = block // execute block
+    // Return to old scope
+    Builder.currentClockAndReset = parentScope
+    res
```

Comment:
Yeah you make a solid point. I decided to return a value because I really like that style of coding. In particular, this would allow people to do things like:

```scala
val myModule = withClockAndReset(someClock, someReset) { Module(new MyModule)) }
```
And thus neatly get a Module with a different clock and/or reset in a single line. It is important that this return value doesn't set people up for failure. @sdtwigg @aswaterman can y'all think of any other reasons why returning a value might be a bad idea? I'm not too too concerned about people trying to return functions but if others think that's reason enough I can switch to returning Unit.